smartsim/entity/ensemble.py

```diff
@@ -298,3 +307,151 @@ def _read_model_parameters(self):
+        present, a number can be passed for specification e.g. "GPU:1".
+
+        Setting ``devices_per_node=N``, with N greater than one will result
+        in the model being stored in the first N devices of type ``device``.
+
+        One of either script (in memory string representation) or script_path (file)
+        must be provided
+
+        :param name: key to store script under
+        :type name: str
+        :param script: TorchScript code
+        :type script: str, optional
+        :param script_path: path to TorchScript code
+        :type script_path: str, optional
+        :param device: device for script execution, defaults to "CPU"
+        :type device: str, optional
+        :param devices_per_node: number of devices on each host
+        :type devices_per_node: int
+        """
+        db_script = DBScript(
+            name=name,
+            script=script,
+            script_path=script_path,
+            device=device,
+            devices_per_node=devices_per_node
+        )
+        self._db_scripts.append(db_script)
+        for entity in self:
+            self._extend_entity_db_scripts(entity, [db_script])
+
+    
+    def add_function(self, name, function=None, device="CPU", devices_per_node=1):
+        """TorchScript function to launch with every entity belonging to this ensemble
+
+        Each script function to the model will be loaded into a
+        non-converged orchestrator prior to the execution
+        of every entity belonging to this ensemble.
+        
+        For converged orchestrators, the :meth:`add_script` method should be used.
+
+        Device selection is either "GPU" or "CPU". If many devices are
+        present, a number can be passed for specification e.g. "GPU:1".
+
+        Setting ``devices_per_node=N``, with N greater than one will result
+        in the model being stored in the first N devices of type ``device``.
+
+        :param name: key to store function under
+        :type name: str
+        :param script: TorchScript code
+        :type script: str, optional
+        :param script_path: path to TorchScript code
+        :type script_path: str, optional
+        :param device: device for script execution, defaults to "CPU"
+        :type device: str, optional
+        :param devices_per_node: number of devices on each host
+        :type devices_per_node: int
+        """
+        db_script = DBScript(
+            name=name,
+            script=function,
+            device=device,
+            devices_per_node=devices_per_node
+        )
+        self._db_scripts.append(db_script)
+        for entity in self:
+            self._extend_entity_db_scripts(entity, [db_script])
+
+    def _extend_entity_db_models(self, model, db_models):
+        entity_db_models = [db_model.name for db_model in model._db_models]
+        for db_model in db_models:
+            if not db_model.name in entity_db_models:
+                model._append_db_model(db_model)
```

Comment:
These new methods are both accessing "protected" methods on `Model`. Given how tightly integrated these two classes are, I am more than willing to call them "friend" classes, but I thought I should bring this to attention to make sure that this is the intended course of action.

Reply:
Yeah, good point. I think they are kinda C++ friends classes in a sense. The other possibility was to "unprotect" the methods, but I think we can leave it like this for the moment being.

Reply:
Definitely going to agree that these methods should not be "unprotected." I don't a user should have access to the method. I like the idea of leaving this as is.